src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -76,54 +76,52 @@ data::InstallationResult SotaUptaneClient::PackageInstall(const Uptane::Target &
-      const std::string correlation_id = pending_target->correlation_id();
+  LOG_INFO << "Pending update for Primary ECU was found, trying to apply it...";
 
-      data::InstallationResult install_res = package_manager_->finalizeInstall(*pending_target);
-      storage->saveEcuInstallationResult(primary_ecu_serial, install_res);
-      if (install_res.success) {
-        storage->saveInstalledVersion(primary_ecu_serial.ToString(), *pending_target,
-                                      InstalledVersionUpdateMode::kCurrent);
-        report_queue->enqueue(
-            std_::make_unique<EcuInstallationCompletedReport>(primary_ecu_serial, correlation_id, true));
-      } else {
-        // finalize failed
-        // unset pending flag so that the rest of the uptane process can
-        // go forward again
-        storage->saveInstalledVersion(primary_ecu_serial.ToString(), *pending_target,
-                                      InstalledVersionUpdateMode::kNone);
-        report_queue->enqueue(
-            std_::make_unique<EcuInstallationCompletedReport>(primary_ecu_serial, correlation_id, false));
-        director_repo.dropTargets(*storage);  // fix for OTA-2587, listen to backend again after end of install
-      }
+  data::InstallationResult install_res = package_manager_->finalizeInstall(*pending_target);
 
-      computeDeviceInstallationResult(nullptr, correlation_id);
-      putManifestSimple();
-    } else {
-      // nothing found on primary
-      LOG_ERROR << "No any pending update for Primary is found";
-    }
+  if (install_res.result_code == data::ResultCode::Numeric::kNeedCompletion) {
+    LOG_INFO << "Pending update for Primary ECU was not applied because reboot was not detected, "
+                "continue with initialization";
+    return;
+  }
+
+  storage->saveEcuInstallationResult(primary_ecu_serial, install_res);
+
+  const std::string correlation_id = pending_target->correlation_id();
+  if (install_res.success) {
+    storage->saveInstalledVersion(primary_ecu_serial.ToString(), *pending_target, InstalledVersionUpdateMode::kCurrent);
+
+    report_queue->enqueue(std_::make_unique<EcuInstallationCompletedReport>(primary_ecu_serial, correlation_id, true));
   } else {
-    LOG_ERROR << "Invalid Uptane metadata in storage.";
+    // finalize failed, unset pending flag so that the rest of the uptane process can go forward again
+    storage->saveInstalledVersion(primary_ecu_serial.ToString(), *pending_target, InstalledVersionUpdateMode::kNone);
+    report_queue->enqueue(std_::make_unique<EcuInstallationCompletedReport>(primary_ecu_serial, correlation_id, false));
   }
 
-  package_manager_->rebootFlagClear();
+  director_repo.dropTargets(*storage);  // fix for OTA-2587, listen to backend again after end of install
+  computeDeviceInstallationResult(nullptr, correlation_id);
+  // Do we really need to send a device manifest here at the initialization phase
```

Comment:
The uptane cycle does not necessarily need to happen at regular intervals. The state of the installation has changed, so we try to send it as soon as it happens.

Reply:
ok